majordodo-core/src/test/java/majordodo/replication/BookkeeperFailuresTest.java

```diff
@@ -0,0 +1,172 @@
+                break;
+            }
+            Thread.sleep(waitForCycle);
+        }
+        assertTrue(ok);
+    }
+        
+    private final LeaderShipChangeListener leaderShiplistener = new LeaderShipChangeListener() {
+        @Override
+        public void leadershipLost() {
+            System.out.println("leadershipLost");
+        }
+        @Override
+        public void leadershipAcquired() {
+            System.out.println("leadershipAcquired");
+        }
+    };
+    
+    private BookKeeper createBookKeeper() throws Exception {
+        byte[] localhostdata = BrokerHostData.formatHostdata(new BrokerHostData(broker1Host, broker1Port, "", false, null));
+        ClientConfiguration config = new ClientConfiguration();
+        config.setThrottleValue(0);
+        
+        ZKClusterManager zkClusterManager = new ZKClusterManager(zkServer.getAddress(), zkServer.getTimeout(), zkServer.getPath(), leaderShiplistener, localhostdata, false);
+        zkClusterManager.waitForConnection();
+        BookKeeper bookKeeper = new BookKeeper(config, zkClusterManager.getZooKeeper());
+        zkClusterManager.start();
+        
+        return bookKeeper;
+    }
+    
+    @Before
+    public void before() {
+        broker1Config.setMaxWorkerIdleTime(1000);
+        broker2Config.setMaxWorkerIdleTime(1000);
+    }
+
+    @Test
+    public void fencingTest() throws Exception {
+        String taskParams = "param";
+
+        Map<String, Integer> tags = new HashMap<>();
+        tags.put(TASKTYPE_MYTYPE, 1);
+
+        assertTrue(broker1.isWritable());
+        
+        long taskId1 = broker1.getClient().submitTask(new AddTaskRequest(0, TASKTYPE_MYTYPE, userId, taskParams, 1, 0, null, 0, null, null)).getTaskId();
+        long taskId2 = broker1.getClient().submitTask(new AddTaskRequest(0, TASKTYPE_MYTYPE, userId, taskParams, 1, 0, null, 0, null, null)).getTaskId();
+        
+        try {
+            long taskId3 = broker2.getClient().submitTask(new AddTaskRequest(0, TASKTYPE_MYTYPE, userId, taskParams, 1, 0, null, 0, null, null)).getTaskId();
+            fail();
+        } catch (Exception ex) {
+            assertEquals("java.lang.Exception: broker_not_leader", ex.getMessage());
+        }
+        
+        ReplicatedCommitLog broker1scl = (ReplicatedCommitLog) broker1.getStatusChangesLog();
+        
+        try (BookKeeper bk = createBookKeeper();) {
+            try (LedgerHandle fenceLedger = bk.openLedger(
+                broker1scl.getCurrentLedgerId(), BookKeeper.DigestType.MAC, "dodo".getBytes(StandardCharsets.UTF_8));) {
+            
+                try {
+                    long taskId4 = broker1.getClient().submitTask(new AddTaskRequest(0, TASKTYPE_MYTYPE, userId, taskParams, 1, 0, null, 0, null, null)).getTaskId();
+                    fail();
+                } catch (LogNotAvailableException ex) {
+                    assertEquals("org.apache.bookkeeper.client.BKException$BKLedgerFencedException", ex.getMessage());
```

Comment:
I think this is test is subject to race conditions,
you have to make the broker stop sending "noops", in fact if the broker issues a "noop" it will detect early the fenced ledger and the state will not be the one you are expecting here